### PR TITLE
Add MAX_RUNTIME option to Torque driver

### LINKED ISSUE #5121: Support MAX_RUNTIME in Torque driver
**Is your feature request related to a problem? Please describe.**
The Torque driver does not support the option MAX_RUNTIME that LSF and Slurm supports

**Describe the solution you'd like**
The same feature as LSF

**Describe alternatives you've considered**
None

### PR DESCRIPTION
**Pull request by kwinkunks**
**Issue**
Resolves #5121 

**Approach**
There is currently a MAX_RUNTIME option for LSF and Slurm, but not for the Torque driver we're using on Azure. This commit adds the option, following the pattern of the recently added TIMEOUT option.

It also adds to test_job_torque.cpp, which tests that the setting and getting machinery works. There is no test for the setting actually being used in the qsub command, or that it actually sets an effective time limit, however.

Also adds documentation of the option.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Updated documentation
- [x] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

**Comment by berland** [comment]
I speculate the LSF driver itself will kill long-running jobs, which makes it clear to ERT why this job failed. Are we able to give a correct error message to the user if we let the queue system implement MAX_RUNTIME on behalf of ERT?

**Comment by kwinkunks** [comment]
@berland Good question.

The function `job_status_type torque_driver_parse_status()` in `torque_driver.cpp` is responsible for parsing the output of `qstat -f`. However, **past** (i.e. completed) jobs do not show up in that output for very long. So our killed jobs may or may not be there.

Even if they are there, jobs that were killed are reported by `qstat -fx` with `job_status` code `F` (finished), so that's no good. They look exactly like properly finished jobs.

However, also in the `qstat -fx` output, killed jobs seem to have `Exit_status` of `-29`, as far as I can tell. [This resource](https://dl.technion.ac.il/docs/cis/public/usg/tamnun/PBS%20exit%20codes.html) suggests it should be `137` so :shrug: 

The most unequivocal evidence that a job was killed is captured in the error file, which the function `torque_driver_get_qstat_status()` seems to call `ert-qstat-err`. If `ert-qstat-std` does not exist, then the error file is reported with `fprintf` with "reading qstat status failed"" --- but really this means that the **job** potentially failed I think, not that qstat failed.

If a job is killed by insufficient `walltime` then its stderr file looks like:

    =>> PBS: job killed: walltime 10 exceeded limit 5

If we wanted to get more information out of that file, I think we'd need to change this code:

```
    if (fs::exists(tmp_std_file)) {
        status = torque_driver_parse_status(tmp_std_file, jobnr_char);
        if (status != JOB_QUEUE_STATUS_FAILURE) {
            unlink(tmp_std_file);
            unlink(tmp_err_file);
        }
    } else
        fprintf(stderr,
                "No such file: %s - reading qstat status failed\n"
                "stderr: %s\n",
                tmp_std_file, tmp_err_file);
    free(tmp_std_file);
    free(tmp_err_file);
```

AFAICT the Slurm driver checks for CANCELLED status and sets `JOB_QUEUE_IS_KILLED`. I guess we're looking for something similar.

**Comment by kwinkunks** [comment]
@kvashchuka I don't quite know what to make of some of these cancelled and failed checks. These are the **Required** ones...

- [ ] Python package / test-linux (3.8, ubuntu-latest, gui-test)
- [ ] Python package / CMake (ubuntu-latest) 
- [ ] Style / check-style (3.10)

`clang-format` says this in CI:

    would reformat src/clib/lib/job_queue/torque_driver.cpp
    would reformat src/clib/old_tests/job_queue/test_job_torque.cpp
    
But I can't get `ert/script/clang-format` to run properly on my machine.

**Comment by xjules** [comment]
The following failed - not sure if that is related though:
```FAILED tests/unit_tests/gui/test_main_window.py::test_that_the_plot_window_contains_the_expected_elements - AssertionError: assert ['default', 'default_0'] == ['default', '..., 'default_3']
  Right contains 3 more items, first extra item: 'default_1'
  Full diff:
  - ['default', 'default_0', 'default_1', 'default_2', 'default_3']
  + ['default', 'default_0']
===== 1 failed, 61 passed, 1340 deselected, 1 warning in 126.35s (0:02:06) =====
```

File: src/clib/lib/job_queue/torque_driver.cpp [hunk 1]
```diff
@@ -317,6 +340,15 @@ stringlist_type *torque_driver_alloc_cmd(torque_driver_type *driver,
         free(resource_string);
     }
 
+    if (driver->max_runtime != 0) {
+        stringlist_append_copy(argv, "-l");
+        char *walltime_string;
+        walltime_string =
+            util_alloc_sprintf("walltime=%d", driver->max_runtime);
```
Comment (xjules) [inline src/clib/lib/job_queue/torque_driver.cpp:347]: I guess that `walltime` was tested - is it in seconds?

Comment (valentin-krasontovitsch) [reply src/clib/lib/job_queue/torque_driver.cpp:347]: little handy habit i learned from prometheus was to add units to your measurement variable names, but i'm not sure if it's applicable / practical here

Comment (kvashchuka) [reply src/clib/lib/job_queue/torque_driver.cpp:347]: yes, it is in seconds, goos idea to add units @valentin-krasontovitsch, thank you! I think I will refactor

**Comment by hnformentin** [comment]
There is some reviewers already, so moved the issue to in review. Hope this is correct :-)

File: docs/reference/configuration/keywords.rst [hunk 2]
```diff
@@ -655,11 +655,14 @@ Keywords controlling the simulations
...
         ::
 
-                -- Let each realizations run for 50 seconds
+                -- Let each realization run for 50 seconds
                 MAX_RUNTIME 50
 
         The MAX_RUNTIME key is optional.
 
+        Note that if you are using a Slurm queue, then you need to pass the option
+        explcitly as a queue option. See :ref:`Slurm's options <slurm_smax_runtime>`.
```
Comment (berland) [inline docs/reference/configuration/keywords.rst:664]: typo explcitly

Comment (kvashchuka) [reply docs/reference/configuration/keywords.rst]: fixed:)

**Comment by sondreso** [comment]
Is this still being worked on? 🙂

**Comment by kwinkunks** [comment]
@sondreso The conclusion was that the feature is already implemented as a 'global' configuration, and the queue option implementation for SLURM should be removed (it can use the global setting). 

It will presumably be implemented in the refactor.

Closing this as redundant.

**Pull request closed without merging**